proxyserver/middleware/versioned_writes.go

```diff
@@ -0,0 +1,549 @@
+	ok, status := v.copyCurrent(writer, request, account, container, versionsContainer, object)
+	if !ok && returnIfStatusError(writer, status) {
+		return
+	}
+	v.next.ServeHTTP(writer, request)
+}
+
+func (v *versionedWrites) handleObject(writer http.ResponseWriter, request *http.Request) {
+	_, account, container, object := getPathParts(request)
+	ctx := GetProxyContext(request)
+	ci, _ := ctx.C.GetContainerInfo(account, container)
+	if ci == nil {
+		v.next.ServeHTTP(writer, request)
+		return
+	}
+
+	versionsContainer := ci.SysMetadata["Versions-Location"]
+	versionsMode := ci.SysMetadata["Versions-Mode"]
+	if versionsMode == "" {
+		versionsMode = "stack"
+	}
+	if versionsContainer == "" {
+		v.next.ServeHTTP(writer, request)
+		return
+	}
+	if unescapedVersionsContainer, err := url.QueryUnescape(versionsContainer); err == nil {
+		versionsContainer = strings.Split(unescapedVersionsContainer, "/")[0]
+	} else {
+		ctx.Logger.Info(fmt.Sprintf("Bad versions container: %s", versionsContainer))
+		v.next.ServeHTTP(writer, request)
+		return
+	}
+
+	if request.Method == "PUT" {
+		v.handleObjectPut(writer, request, account, container, versionsContainer, object)
+	} else if versionsMode == "history" {
+		v.handleObjectDeleteHistory(writer, request, account, container, versionsContainer, object)
+	} else {
+		v.handleObjectDeleteStack(writer, request, account, container, versionsContainer, object)
+	}
+}
+
+func returnIfStatusError(writer http.ResponseWriter, status int) bool {
+	if status/100 == 2 {
+		return false
+	}
+	if status == 403 {
+		srv.StandardResponse(writer, 403)
+	} else if status/100 == 4 {
+		srv.StandardResponse(writer, 412)
+	} else {
+		srv.StandardResponse(writer, 500)
+	}
+	return true
+}
+
+func (v *versionedWrites) ServeHTTP(writer http.ResponseWriter, request *http.Request) {
+	apiReq, account, container, object := getPathParts(request)
+	if container == "" && object == "" {
+		v.next.ServeHTTP(writer, request)
+		return
+	}
+	if !apiReq || account == "" {
+		v.next.ServeHTTP(writer, request)
+		return
+	}
+	if GetProxyContext(request).Source == "VW" {
+		v.next.ServeHTTP(writer, request)
+		return
+	}
+
+	if !v.enabled {
+		// Can't set these headers if disabled
+		if (request.Method == "POST" || request.Method == "PUT") &&
+			(request.Header.Get(CLIENT_VERSIONS_LOC) != "" ||
+				request.Header.Get(CLIENT_HISTORY_LOC) != "") {
+
+			srv.SimpleErrorResponse(writer, 412, "Versioned Writes is disabled")
+			return
+		}
+		v.next.ServeHTTP(writer, request)
+		return
+	}
+
+	if object == "" {
+		v.handleContainer(writer, request)
+	} else if request.Method == "PUT" || request.Method == "DELETE" {
+		v.handleObject(writer, request)
+	} else {
+		v.next.ServeHTTP(writer, request)
+	}
+}
+
+func NewVersionedWrites(config conf.Section, metricsScope tally.Scope) (func(http.Handler) http.Handler, error) {
+	RegisterInfo("versioned_writes", map[string]interface{}{"allowed_flags": []string{CLIENT_VERSIONS_LOC, CLIENT_HISTORY_LOC}})
```

Comment:
it might sound silly but we prob do need to strings.ToLower these. if somebody's checking for something in json payload they probably will just look for the lower case key